canjam/inboundworker.py

```diff
@@ -0,0 +1,82 @@
+from queue import Queue
+from threading import Thread
+
+from canjam.jamsocket import Jamsocket
+from canjam.message import (
+    Message,
+    Sound,
+    ReqUserList,
+    RspUserList,
+    NewUser,
+    DelUser,
+    Die,
+)
+from canjam.user import User
+
+
+class InboundWorker:
+    """
+    A multi-threaded worker that listens for incoming messages on a Jamsocket and
+    processes them. The worker maintains a list of users and a queue of incoming
+    Sound messages. The worker can be started and stopped with the start() and
+    stop() methods, or by using with...as syntax.
+    """
+
+    sock: Jamsocket
+    in_queue: Queue[Message]
+    user_list: list[User]
+    __worker_thread: Thread
+
+    def __init__(
+        self,
+        sock: Jamsocket,
+        in_queue: Queue[Message] = Queue(),
+        user_list: list[User] = [],
+    ):
+        self.sock = sock
+        self.in_queue = in_queue
+        self.user_list = user_list
+        self.__worker_thread = Thread(target=self.__worker_job)
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *_):
+        self.stop()
+
+    def __worker_job(self):
+        while True:
+            data, address = self.sock.recvfrom()
+            try:
+                message = Message.deserialize(data)
+            except:
+                print("Failed to deserialize message from", address)
+                continue
+            match message:
+                case ReqUserList():
+                    rsp = RspUserList(self.user_list)
+                    # TODO: handle this error case... remove user if they don't respond to the message?
+                    assert self.sock.sendto_reliably(rsp.serialize(), address)
```

Comment:
If they don't respond to the message, we could send them a `Die` message to tell their whole program to shut down (i.e. if they're hanging on a socket or something else weird)